### PR TITLE
Global Styles: Prevent duplicate CSS for block style variations

### PR DESCRIPTION
**Pull request by aaronrobertshaw**
This PR backports the PHP updates from: https://github.com/WordPress/gutenberg/pull/62465

It now also includes the missing changes to `get_block_nodes` from https://github.com/WordPress/gutenberg/pull/41217.

The changes prevent redundant or duplicate CSS being generated for block style variations.

### Test Instructions

1. Register a custom block style variation e.g. through a theme.json partial: `/styles/partial.json`

```
{
	"$schema": "https://schemas.wp.org/trunk/theme.json",
	"version": 2,
	"title": "Custom",
	"slug": "custom",
	"blockTypes": [ "core/group" ],
	"styles": {
		"color": {
			"background": "aliceblue"
		}
	}
}
```

2. In both editors add a group block and apply the `Custom` block style variation.
3. Inspect the group block and confirm there is only a single `is-style-custom-*` css class matched
4. Save the page or post and confirm the same on the frontend

Trac ticket: https://core.trac.wordpress.org/ticket/61443

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

### CODE REVIEW

**Comment by aaronrobertshaw** [comment]
I noticed while creating this backport of https://github.com/WordPress/gutenberg/pull/62465 that there's a difference in signature between core and Gutenberg for `get_block_nodes`.

Before this change:

Core: `private static function get_block_nodes( $theme_json )`
Gutenberg: `private static function get_block_nodes( $theme_json, $selectors = array() )`

I'll try and sort this out in time for beta 3.

**Comment by aaronrobertshaw** [comment]
The PR modifying the function signature in Gutenberg was https://github.com/WordPress/gutenberg/pull/41217.

From https://github.com/WordPress/gutenberg/pull/41217#issuecomment-1354426105 it looks like it the fact the PR wasn't backported was already known however perhaps not that "everything will be fine" as suggested after https://github.com/WordPress/gutenberg/pull/46579.

As far as I can tell, I don't think there'll be any harm in updating `get_block_nodes` function to match Gutenberg's version. I'll make that update shortly.

**Comment by aaronrobertshaw** [comment]
I've made the required updates from https://github.com/WordPress/gutenberg/pull/41217.

Flagging this as ready for review once the Gutenberg PR lands.

File: src/wp-includes/class-wp-theme-json.php [hunk 1]
```diff
@@ -1260,6 +1260,7 @@ public function get_settings() {
 	 *     @type string $scope                   Makes sure all style are scoped to a given selector
 	 *     @type string $root_selector           Overwrites and forces a given selector to be used on the root node
 	 *     @type bool   $skip_root_layout_styles Omits root layout styles from the generated stylesheet. Default false.
+	 *     @type bool   $block_style_variations  Includes styles for block style variations in the generated stylesheet. Default false.
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-theme-json.php]: Include this new type in line. 1250

File: src/wp-includes/class-wp-theme-json.php [hunk 2]
```diff
@@ -2564,12 +2571,19 @@ private static function update_separator_declarations( $declarations ) {
...
+	 * @param array $options {
+	 *     Optional. An array of options for now used for internal purposes only (may change without notice).
+	 *
+	 *     @type bool   $block_style_variations  Includes nodes for block style variations. Default false.
+	 * }
 	 * @return array The block nodes in theme.json.
 	 */
-	private static function get_block_nodes( $theme_json ) {
-		$selectors = static::get_blocks_metadata();
+	private static function get_block_nodes( $theme_json, $selectors = array(), $options = array() ) {
```
Comment (ramonjd) [inline src/wp-includes/class-wp-theme-json.php:2586]: I've checked these changes against https://github.com/WordPress/gutenberg/pull/41217 and smoke tested this PR.

LGTM

Thanks for cleaning that up.

**Comment by ramonjd** [comment]
This PR is testing well for me. For blocks with the block style variation applied, I'm seeing the following, related classnames in the class attribute in both the editor and frontend:

`is-style-custom is-style-custom-b321ad1e-3e38-4a3c-af91-ef82f0c9855d`

**Review by ramonjd — APPROVED** [review]

File: src/wp-includes/class-wp-theme-json.php (hunk 1 above)
Comment (aaronrobertshaw) [reply src/wp-includes/class-wp-theme-json.php]: Done: https://github.com/WordPress/wordpress-develop/pull/6827/commits/7a72f569f820b706d0bae5837d043e002eb07933

**Comment by aaronrobertshaw** [comment]
> I'm seeing the following, related classnames in the class attribute in both the editor and frontend

For the moment, this is expected. The `is-style-custom` classname is what's applied by the editor and is hooked onto by the block support to generate the class specific to this instance or application of the block style, `is-style-custom--<hash>`.

File: src/wp-includes/class-wp-theme-json.php [hunk 3]
```diff
@@ -2446,12 +2447,18 @@ protected static function get_setting_nodes( $theme_json, $selectors = array() )
...
 	 *
 	 * @since 5.8.0
+	 * @since 6.6.0 Added options array for modifying generated nodes.
 	 *
 	 * @param array $theme_json The tree to extract style nodes from.
 	 * @param array $selectors  List of selectors per block.
+	 * @param array $options {
+	 *     Optional. An array of options for now used for internal purposes only (may change without notice).
+	 *
+	 *     @type bool   $block_style_variations  Includes style nodes for block style variations. Default false.
```
Comment (mukeshpanchal27) [inline src/wp-includes/class-wp-theme-json.php]:
```suggestion
	 *     @type bool $block_style_variations Includes style nodes for block style variations. Default false.
```
Spaces not needed here.

**Comment by tellthemachines** [comment]
Just a quick note that applying this patch on top of the latest trunk (after the [custom CSS fix](https://core.trac.wordpress.org/changeset/58421)) still shows everything working as expected 👍

**Review by oandregal — APPROVED** [review]

**Comment by oandregal** [comment]
Committed at https://core.trac.wordpress.org/changeset/58422

**Pull request closed without merging**